libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,558 @@
+/*
+** 2024-07-04
+**
+** Copyright 2024 the libSQL authors
+**
+** Permission is hereby granted, free of charge, to any person obtaining a copy of
+** this software and associated documentation files (the "Software"), to deal in
+** the Software without restriction, including without limitation the rights to
+** use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+** the Software, and to permit persons to whom the Software is furnished to do so,
+** subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in all
+** copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+** FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+** COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+** IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+** CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+**
+******************************************************************************
+**
+** libSQL basic vector functions
+*/
+#ifndef SQLITE_OMIT_VECTOR
+
+#include "sqliteInt.h"
+#include "vectorInt.h"
+
+#define MAX_FLOAT_CHAR_SZ  1024
+
+/**************************************************************************
+** Utility routines for dealing with Vector objects
+**************************************************************************/
+
+size_t vectorDataSize(VectorType type, VectorDims dims){
+  switch( type ){
+    case VECTOR_TYPE_FLOAT32:
+      return dims * sizeof(float);
+    case VECTOR_TYPE_FLOAT64:
+      return dims * sizeof(double);
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+void vectorInit(Vector *pVector, VectorType type, VectorDims dims, void *data){
+  pVector->type = type;
+  pVector->dims = dims;
+  pVector->data = data;
+  pVector->flags = 0;
+}
+
+/*
+ * Allocate a Vector object and its data buffer
+*/
+Vector *vectorAlloc(VectorType type, VectorDims dims){
+  void *pVector = sqlite3_malloc(sizeof(Vector) + vectorDataSize(type, dims));
+  if( pVector==NULL ){
+    return NULL;
+  }
+  vectorInit(pVector, type, dims, ((char*) pVector) + sizeof(Vector));
+  return pVector;
+}
+
+/*
+ * Allocate a Vector object and its data buffer from the SQLite context. 
+*/
+static Vector* vectorContextAlloc(sqlite3_context *context, int type, int dims){
+  void *pVector = sqlite3_malloc64(sizeof(Vector) + vectorDataSize(type, dims));
+  if( pVector==NULL ){
+    sqlite3_result_error_nomem(context);
+    return NULL;
+  }
+  vectorInit(pVector, type, dims, ((char*) pVector) + sizeof(Vector));
+  return pVector;
+}
+
+/*
+ * Free a Vector object and its data buffer allocated, unless the vector is static.
+*/
+void vectorFree(Vector *pVector){
+  if( pVector==0 ){
```

Comment:
Why does one function use NULL and another 0? Is there a reason for this inconsistency?

Reply:
No reason - I guess we better to explicitly use NULL everywhere. Will fix that